File: lld/week-02-03-patterns/factory/go/factory.go
// Package factory implements the Factory Method pattern: a
// NotificationFactory that produces Email/SMS/Push notifications behind
// a common Notification interface, so callers never construct concrete
// types directly.
package factory

import (
	"errors"
	"fmt"
)

type NotificationType int

const (
	Email NotificationType = iota
	SMS
	Push
)

type Notification interface {
	Send(recipient, message string) string
}

type EmailNotification struct{}

func (EmailNotification) Send(recipient, message string) string {
	return fmt.Sprintf("Email to %s: %s", recipient, message)
}

type SMSNotification struct{}

func (SMSNotification) Send(recipient, message string) string {
	return fmt.Sprintf("SMS to %s: %s", recipient, message)
}

type PushNotification struct{}

func (PushNotification) Send(recipient, message string) string {
	return fmt.Sprintf("Push to %s: %s", recipient, message)
}

var ErrUnknownNotificationType = errors.New("unknown notification type")

// CreateNotification is the factory method: it hides construction of the
// concrete Notification type behind a single switch, so adding a new
// channel means adding one case here rather than touching every call site.
func CreateNotification(t NotificationType) (Notification, error) {
	switch t {
	case Email:
		return EmailNotification{}, nil
	case SMS:
		return SMSNotification{}, nil
	case Push:
		return PushNotification{}, nil
	default:
		return nil, ErrUnknownNotificationType
	}
}

File: lld/week-02-03-patterns/factory/python/factory.py
"""Factory Method LLD — a NotificationFactory producing Email/SMS/Push
notifications behind a common Notification protocol. See ../README.md for
the design writeup.
"""
from __future__ import annotations

from enum import Enum, auto
from typing import Protocol


class NotificationType(Enum):
    EMAIL = auto()
    SMS = auto()
    PUSH = auto()


class Notification(Protocol):
    def send(self, recipient: str, message: str) -> str: ...


class EmailNotification:
    def send(self, recipient: str, message: str) -> str:
        return f"Email to {recipient}: {message}"


class SMSNotification:
    def send(self, recipient: str, message: str) -> str:
        return f"SMS to {recipient}: {message}"


class PushNotification:
    def send(self, recipient: str, message: str) -> str:
        return f"Push to {recipient}: {message}"


class UnknownNotificationTypeError(Exception):
    pass


_REGISTRY = {
    NotificationType.EMAIL: EmailNotification,
    NotificationType.SMS: SMSNotification,
    NotificationType.PUSH: PushNotification,
}


def create_notification(notification_type: NotificationType) -> Notification:
    cls = _REGISTRY.get(notification_type)
    if cls is None:
        raise UnknownNotificationTypeError(f"unknown notification type: {notification_type}")
    return cls()


def _demo() -> None:
    for nt in (NotificationType.EMAIL, NotificationType.SMS, NotificationType.PUSH):
        notification = create_notification(nt)
        print(notification.send("user@example.com", "your order has shipped"))


if __name__ == "__main__":
    _demo()
